numba/np/arrayobj.py

```diff
@@ -1389,6 +1408,16 @@ def impl(array, shape):
     elif isinstance(shape, types.Integer):
         def impl(array, shape):
             return np.broadcast_to(array, (shape,))
+    elif isinstance(shape, types.Tuple) and shape.count == 0:
+        is_scalar_array = isinstance(array, types.Array) and array.ndim == 0
+        if type_is_scalar(array) or is_scalar_array:
```

Comment:
Perhaps add a comment here saying that `array` must be supported by `type_can_asarray` as checked above. I just spent a few minutes trying to work out why inputs such as `'hello'` are commented in the test below, it's because whilst they are permitted scalar types, they fail the `type_can_asarray` check so cannot reach here.

This also lead to the discovery of this bug:
```python
from numba import njit
import numpy as np


@njit
def foo():
    return np.broadcast_to('a', (2, 3))

print(foo())
```
During type inference, for the non-literal type attempt, this function correctly fails to accept `'a'` as an input (as it's a `unicode_type`). For the subsequent literal type attempt, the types are: `(Literal[str](a), Tuple(Literal[int](2), Literal[int](3)))`. The literal string 'a', is correctly handled and accepted, but the shape tuple is now a `types.Tuple` of literal integers (i.e. heterogeneous), and this is not handled correctly and leads to calling the `else:` branch fall through below.

Reply:
Hi Stuart, I've addressed the issue.